### PR TITLE
Open62541 - Allow more clang versions

### PR DESCRIPTION
**Pull request by Kischy**
Specify library name and version:  **open62541/1.3.8**

Allows a broader variety of clang compiler versions. See issue https://github.com/conan-io/conan-center-index/issues/20579

---

- [x] I've read the [contributing guidelines](https://github.com/conan-io/conan-center-index/blob/master/CONTRIBUTING.md).
- [x] I've used a [recent](https://github.com/conan-io/conan/releases/latest) Conan client version close to the [currently deployed](https://github.com/conan-io/conan-center-index/blob/master/.c3i/config_v1.yml#L6).
- [ ] I've tried at least one configuration locally with the [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

File: recipes/open62541/all/conanfile.py
Comment (Ahajha) [inline recipes/open62541/all/conanfile.py:221]: Do you know if it's just 1.3.8 onwards that adds support for newer clang versions? From what you said in the issue, I'd guess that if it's still relevant at all, it's only for much older versions.

File: recipes/open62541/all/test_package/CMakeLists.txt
Comment (Ahajha) [inline recipes/open62541/all/test_package/CMakeLists.txt:37]: Usually test packages are just testing for building/linking/making sure the package works. I don't know if there's a precedent for adding sanitizers to a test package.

File: recipes/open62541/all/conanfile.py
Comment (Kischy) [reply recipes/open62541/all/conanfile.py:221]: My thinking was that this way the package does not change for older versions, i.e. it is "backwards compatible". 

In generell I see no reason to only allow versions >= 1.3.8. We could just remove this check and see what happens. I just do not want to debug old versions of the package.

File: recipes/open62541/all/test_package/CMakeLists.txt
Comment (Kischy) [reply recipes/open62541/all/test_package/CMakeLists.txt:37]: Do be honest, I do not know why this is necessary in the first place. If I do not add this linker option I get linker errors like these ones:
```
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __ubsan_handle_out_of_bounds [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __ubsan_handle_pointer_overflow [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __ubsan_handle_type_mismatch_v1 [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __ubsan_handle_nonnull_arg [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __ubsan_handle_divrem_overflow [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __ubsan_handle_sub_overflow [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __ubsan_handle_add_overflow [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __ubsan_handle_mul_overflow [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __ubsan_handle_shift_out_of_bounds [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __ubsan_handle_load_invalid_value [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __sanitizer_cov_trace_pc_guard [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __sanitizer_cov_trace_pc_guard_init [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __asan_register_globals [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __asan_unregister_globals [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __asan_init [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __asan_version_mismatch_check_v8 [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __asan_report_load_n [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __asan_loadN [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __asan_report_load1 [--no-allow-shlib-undefined]
ld: error: /home/conan/w/prod-v1/bsr/48966/cdbab/.conan/data/open62541/1.3.8/_/_/package/196f1131654c20996501278721a228f23b2dcd95/lib/libopen62541.so: undefined reference to __asan_report_load2 [--no-allow-shlib-undefined]
```

See https://c3i.jfrog.io/c3i/misc/logs/pr/20892/10-linux-clang/open62541/1.3.8//196f1131654c20996501278721a228f23b2dcd95-test.txt

Do you understand why these references are not found by the linker?

File: recipes/open62541/all/conanfile.py
Comment (Ahajha) [reply recipes/open62541/all/conanfile.py:221]: Adding support for more versions of clang only adds users, so it's backwards compatible as is, but it's preferable to have as many valid configurations as possible.

I can check a few clang versions locally and see what happens. But also, I think some of the patch revisions can be removed, currently there's:
```
1.3.8
1.3.6
1.3.4
1.3.1
1.2.6
1.2.4
1.2.2
1.1.6
1.1.5
1.1.3
1.0.6
1.0.3
```
I don't know how much changes between versions in open62541, but I would think that shortening the list to just the latest patch version in each minor version would be sufficient:
```
1.3.8
1.2.6
1.1.6
1.0.6
```
@RubenRBS Thoughts on this?

Comment (Kischy) [reply recipes/open62541/all/conanfile.py:221]: Ok, if we can remove versions I am all for it. It takes for ever to build all of these packages during a PR.

File: recipes/open62541/all/test_package/CMakeLists.txt
Comment (Ahajha) [reply recipes/open62541/all/test_package/CMakeLists.txt:37]: Interesting. Do you know if 1.3.7/8 adds this to their build by default? If so, while I find it a little odd, it would mean that you need to add `-fsanitize=address` to the package's build and link flags in this version and up. We don't want consumers to have to have a weird link error when they try to use the package that they then have to fix, it should work out of the box.

Comment (Ahajha) [reply recipes/open62541/all/test_package/CMakeLists.txt:37]: I think that would be reasonable. I imagine it's there for debugging of their code, so I don't know if it's that useful for use users of the package. That said, we could expose it as an option, which would conditionally set that variable and add those flags.

Comment (Kischy) [reply recipes/open62541/all/test_package/CMakeLists.txt]: Ok I exposed the option and changed the test_package back to the original.

File: recipes/open62541/all/conanfile.py
Comment (Kischy) [reply recipes/open62541/all/conanfile.py]: I already made this change (I really want to be done with this). If this change is not needed we can always revert to commit `21c656d91905b5ebe971790e5c1d2f4d8f97f969`.

File: recipes/open62541/all/test_package/CMakeLists.txt
Comment (Ahajha) [reply recipes/open62541/all/test_package/CMakeLists.txt]: Looks like you also need to add a couple other things:
1. Pass the option from Conan to CMake through the CMake toolchain in `generate()`. Something like this existing option: `tc.variables["UA_ENABLE_SUBSCRIPTIONS"] = self.options.subscription != False`
2. If the option is enabled, modify the `package_info()` to add the sanitizer flags.

File: recipes/open62541/all/conanfile.py
Comment (Ahajha) [reply recipes/open62541/all/conanfile.py]: I'm having difficult building the package at all locally, I'll try to test those clang versions tomorrow.

File: recipes/open62541/all/test_package/CMakeLists.txt
Comment (Kischy) [reply recipes/open62541/all/test_package/CMakeLists.txt]: Arg, yeah, that makes sense...

Comment (Kischy) [reply recipes/open62541/all/test_package/CMakeLists.txt]: Ok, the first part makes sense. But why should I enable the sanitizer flags in package_info? This would mean I have to duplicate this check from the cmake of the original open62541 Repo
https://github.com/open62541/open62541/blob/7837c1d6246fa78bba06d5e65ae82579867a1d5c/CMakeLists.txt#L652
and add all the sanitizer flags set in this line
https://github.com/open62541/open62541/blob/7837c1d6246fa78bba06d5e65ae82579867a1d5c/CMakeLists.txt#L657
.

That sounds odd to me.

**Comment by Kischy** [comment]
@Ahajha Is it possible to cancel the CI for older commits?

File: recipes/open62541/all/test_package/CMakeLists.txt
Comment (Kischy) [reply recipes/open62541/all/test_package/CMakeLists.txt]: I just realized that `UA_ENABLE_SUBSCRIPTIONS` is not available in release 1.3.8. I will add a patch for this version that removes the sanitizers.

**Comment by Ahajha** [comment]
> @Ahajha Is it possible to cancel the CI for older commits?

I believe they get cancelled automatically

File: recipes/open62541/all/conanfile.py
Comment (Ahajha) [reply recipes/open62541/all/conanfile.py]: I can't seem to build in any configuration locally, unsure what the issue is. Having those clang checks for only later versions is probably fine for now.

Comment (Kischy) [reply recipes/open62541/all/conanfile.py]: It should work now. I used a patch to get rid of the sanitizer.

File: recipes/open62541/all/conanfile.py
Comment (Kischy) [reply recipes/open62541/all/conanfile.py]: Implemented. I do not know how to stop the CI from running.

File: recipes/open62541/all/patches/0001-disable-sanitizers-1_3_x.patch
Comment (Kischy) [reply recipes/open62541/all/patches/0001-disable-sanitizers-1_3_x.patch]: Implemented. I do not know how to stop the CI from running.

**Comment by Kischy** [comment]
@Ahajha Will this be squash merged? The history is not needed imho.

**Comment by Ahajha** [comment]
> @Ahajha Will this be squash merged? The history is not needed imho.

Yes, the commits are squashed.

**Comment by Kischy** [comment]
@Ahajha Who do I need to ask for a second review?

**Comment by Ahajha** [comment]
> @Ahajha Who do I need to ask for a second review?

Hopefully Ruben will get to it soon, I think he's been a bit busy recently. Also my approval "doesn't count" so to speak, I don't have write access.

**Comment by Kischy** [comment]
> > @Ahajha Who do I need to ask for a second review?
> 
> Hopefully Ruben will get to it soon, I think he's been a bit busy recently. Also my approval "doesn't count" so to speak, I don't have write access.

Ok, so I still need a second person with write access to review this...

**Comment by Kischy** [comment]
@Ahajha, @RubenRBS Do you have the possibility to link issue https://github.com/conan-io/conan-center-index/issues/20579 to this PR? I do not have the rights to do this.

**Comment by Kischy** [comment]
@RubenRBS Can I help you in any way to get the reviews faster? :)

**Comment by Kischy** [comment]
@Ahajha Hmm, nothing happening here. How can I motivate someone with write access to review this PR?

**Comment by Ahajha** [comment]
> @Ahajha Hmm, nothing happening here. How can I motivate someone with write access to review this PR? 

Tbh, I wish I knew. I've had a PR sitting for a few weeks now, struggling to get a review myself.

**Comment by Kischy** [comment]
> > @Ahajha Hmm, nothing happening here. How can I motivate someone with write access to review this PR? 
> 
> Tbh, I wish I knew. I've had a PR sitting for a few weeks now, struggling to get a review myself.

Arg... that's depressing. This is blocking other work for me. If this takes so long I will have to find a different solution. 😢

**Comment by Kischy** [comment]
@Ahajha Which one is your PR? I will try something that might make me quite unpopular...

**Comment by Ahajha** [comment]
Please don't. Regardless, it's a weekend. On Monday I'll try to ask in Slack (Cpplang) for some reviews.

**Comment by Kischy** [comment]
> Please don't. Regardless, it's a weekend. On Monday I'll try to ask in Slack (Cpplang) for some reviews.

Ok , then I wait. Thank you.

**Review by uilianries — COMMENTED** [review]
LGTM

**Review by RubenRBS — APPROVED** [review]
Thanks a lot for your patience, after looking into the initial reasons of why the version limits, this seems to be ok :)

**Comment by RubenRBS** [comment]
> Please don't. Regardless, it's a weekend. On Monday I'll try to ask in Slack (Cpplang) for some reviews.

> Ok , then I wait. Thank you.

Thank you for your understanding @Kischy @Ahajha

**Review by uilianries — APPROVED** [review]

**Comment by Kischy** [comment]
@RubenRBS  @uilianries @Ahajha  Thank you so much :)

**Pull request merged**
